pyam/_aggregate.py

```diff
@@ -42,76 +52,62 @@ def _aggregate(df, variable, components=None, method=np.sum):
                 mapping[c] = v
 
     # rename all components to `variable` and aggregate
-    _df = df.data[df._apply_filters(variable=mapping.keys())].copy()
-    _df["variable"].replace(mapping, inplace=True)
+    _df = df._data[df._apply_filters(variable=mapping.keys())]
+    _df.index = replace_index_values(_df, "variable", mapping)
     return _group_and_agg(_df, [], method)
 
 
-def _aggregate_recursive(df, variable, method=np.sum):
+def _aggregate_recursive(df, variable):
     """Recursive aggregation along the variable tree"""
-    _df_aggregated = None
-    _df = df.copy()
-
-    # iterate over variables to find all subcategories to be aggregated
-    sub_variables = []
-    for d in reversed(range(1, max(find_depth(df.data.variable)) + 1)):
-        depth = find_depth(df.data.variable)
-        var_list = df.data.variable[[i == d for i in depth]].unique()
-        vars_up = pd.Series([reduce_hierarchy(i, -1) for i in var_list]).unique()
-
-        if [i for i, entr in enumerate(vars_up) if entr.startswith(variable)]:
-            for v in vars_up:
-                sub_variables.append(v)
 
-    sub_variables = reversed(sorted(set(sub_variables)))
+    # downselect to components of `variable`, initialize list for aggregated (new) data
+    _df = df.filter(variable=f"{variable}|*")
+    data_list = []
 
-    # iterate over subcategories (bottom-up) and perform aggregation
-    for entry in sub_variables:
-        _df.aggregate(variable=entry, append=True)
-        _df_temp = _df.aggregate(variable=entry, append=False)
+    # iterate over variables (bottom-up) and aggregate all components up to `variable`
+    for d in reversed(range(find_depth(variable), max(find_depth(_df.variable)))):
+        components = compress(_df.variable, find_depth(_df.variable, level=d + 1))
+        var_list = set([reduce_hierarchy(v, -1) for v in components])
 
-        if _df_aggregated is None:
-            _df_aggregated = _df_temp.copy()
-        else:
-            _df_aggregated.append(_df_temp, inplace=True)
+        # a temporary dataframe allows to distinguish between full data and new data
+        temp_df = _df.aggregate(variable=var_list)
+        _df.append(temp_df, inplace=True)
+        data_list.append(temp_df._data)
 
-    return _df_aggregated.data
+    return pd.concat(data_list)
```

Comment:
I'm pretty sure that re-implementation fixes also bugs that I had with recursive aggregation previously where it would sometimes only perform the lowest level aggregation.